contributors/design-proposals/loadbalancer-proposal.md

```diff
@@ -0,0 +1,396 @@
+	LoadBalancerPending LoadBalancerPhase = "Running"
+	// used for LoadBalancers that failed to be correctly recycled or deleted after being released from a claim
+	LoadBalanceFailed LoadBalancerPhase = "Failed"
+)
+
+// LoadBalancerIngress represents the status of a load-balancer ingress point:
+// traffic should be sent to an ingress point.
+type LoadBalancerIngress struct {
+	// IP is set for load-balancer ingress points that are IP based
+	// (typically GCE or OpenStack load-balancers)
+	IP string `json:"ip,omitempty"`
+
+	// Hostname is set for load-balancer ingress points that are DNS based
+	// (typically AWS load-balancers)
+	Hostname string `json:"hostname,omitempty"`
+}
+```
+
+### LoadbalancerClaim API
+
+```go
+type LoadbalancerClaim struct {
+	unversioned.TypeMeta `json:",inline"`
+	ObjectMeta           `json:"metadata,omitempty"`
+
+	// Spec defines a loadbalancer owned by the cluster
+	Spec LoadBalancerClaimSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about loadbalancer.
+	Status LoadBalancerClaimStatus `json:"status,omitempty"`
+}
+
+type LoadBalancerClaimSpec struct {
+	// A label query over loadbalancer to consider for binding. This selector is
+	// ignored when LoadBalancerName is set
+	Selector *unversioned.LabelSelector `json:"selector,omitempty"`
+	// Resources represents the minimum resources required
+	Resources ResourceRequirements `json:"resources,omitempty"`
+	// LoadBalancerName is the binding reference to the LoadBalancerName backing this
+	// claim. When set to non-empty value Selector is not evaluated
+	LoadBalancerName string `json:"loadbalancerName,omitempty"`
+}
+
+type LoadBalancerClaimStatus struct {
+}
+```
+
+### Ingress API Changes
+
+Add a new knob "LoadBalancer" in IngressSpec, so that Ingress can "use" or
+"consume" a LoadBalancer, either directly or through LoadBalancerClaim.
+
+```go
+type IngressSpec struct {
+    /*
+    ...
+    */
+
+    LoadBalancer LBSource `json:"loadBalancer"`
+}
+
+type LBSource struct {
+    GCELoadBalancer *GCELoadBalancerSource `json:"gceLoadBalancer,omitempty"`
+	AWSLoadBalancer *AWSLoadBalancerSource `json:"awsLoadBalancer,omitempty"`
+	BareMetalLoadBalancer *BareMetalLoadBalancerSource `json:"bareMetalLoadBalancer,omitempty"`
+	LoadBalancerClaim *LoadBalancerClaimSource `json:"loadBalancerClaim,omitempty"`
+}
+
+type LoadBalancerClaimSource struct {
+	// ClaimName is the name of a LoadBalancerClaim in the same namespace as the ingress using this lb
+	ClaimName string `json:"claimName"`
+}
+```
+
+## Implementation Detail
+
+### loadbalancer
+
+The current loadbalancer (ingress-controller) implementation has some limitations
+when we have multiple loadbalancers:
+
+* we need to deploy multiple ingress-controller pod even on cloud provider, it
+  results in some excessive use of resource. Actually it fine that a cluster
+  just has one process which list&watch all ingresses, and call cloud provider
```

Comment:
Reading this, it seems that the goal of this design is to reduce the number of list&watch on apiserver. Instead of all ingress controllers doing their list&watch, the proposed loadbalancer controller will be the only one performing list&watch.

Is that true? If yes then we should state that goal explicitly at top.